### PR TITLE
feat(llm): retain Anthropic/Gemini state and portable plain-text reasoning

### PR DESCRIPTION
**Pull request by furgalep**
## Summary

Retain opaque provider state for compatible replay **and persist/replay plain-text reasoning across models**. Readable reasoning is also exported through OTLP; encrypted state and signatures are not.

- extend #310's Chat envelope to Anthropic signed/redacted thinking and Gemini thought signatures
- preserve provider-readable reasoning text on canonical `LLMResponse.reasoning`
- restore opaque state only when provider, API style, exact model, and complete public carrier match
- keep Gemini's private inline tool-call IDs out of NOOA's public tool API, then restore both call and result IDs at dispatch
- strip direct provider-private fields that did not arrive through a validated envelope
- warn on expected compatibility fallback and hard-fail malformed current state
- redact opaque state from HTTP, journal, tracing, and flattened diagnostics

## Why this layer exists

A provider response has public conversational meaning and sometimes private protocol state. Anthropic returns signed thinking blocks; Gemini returns thought signatures, including signatures embedded by LiteLLM in tool-call IDs. Those artifacts may be required for the next compatible request, but they are not ordinary prompt text and must never reach another provider.

NOOA therefore needs two representations at two different boundaries:

- the durable, provider-independent `LLMResponse` used by events, SQLite sessions, formatters, middleware, and strategies;
- the provider-native message reconstructed only at the final UnifiedLLM dispatch edge.

The event record stores readable reasoning as text and the provider artifact as one opaque `llm_state` blob. It does not grow Anthropic/Gemini fields. The replay adapter opens that blob only after the target scope and the unchanged public assistant carrier prove that it belongs.

This also protects cache stability. Reconstructing the original provider message at the edge preserves the provider's expected turn shape; the public event history remains canonical and does not accumulate synthetic or partially translated provider messages.

## Code walkthrough: what changed and why

1. **Anthropic capture and replay.**  
   **What:** CompletionClient captures native thinking and redacted-thinking blocks, including signatures, and restores them only on a matching Anthropic Chat scope and unchanged carrier.  
   **Why:** signed thinking must remain paired with the exact turn that produced it and is unsafe outside that compatibility boundary.

2. **Gemini capture and replay.**  
   **What:** UnifiedLLM captures message-level and per-tool thought signatures. When provider evidence identifies LiteLLM's inline signature suffix, NOOA exposes a clean public tool-call ID, stores the suffix privately, and restores the private ID on both the assistant tool call and its later tool-result message.  
   **Why:** Gemini requires matching signed call/result IDs, while NOOA tools, event storage, and other providers need stable public IDs with no opaque payload embedded in them. A non-Gemini ID merely containing `__thought__` remains an ordinary ID, even with an empty suffix.

3. **Shared full-carrier proof.**  
   **What:** #310's compact SHA-256 carrier binding is generalized across supported Chat state: assistant text plus each ordered tool call's ID, name, and exact arguments must still match. Gemini public-to-wire IDs are recorded while restoring signatures rather than recovered from a duplicated stored carrier. Signature count is checked against the current fingerprint-matched public calls.  
   **Why:** checking only IDs or list length could attach stale signatures after middleware changes the actual request.

4. **Plain-text reasoning: persistence, cross-model replay, and OTLP export.**  
   **What:** Chat reasoning and OpenAI Responses summaries populate `LLMResponse.reasoning`. When opaque replay is unavailable, #312's provider-independent path demotes that text to ordinary assistant context. The tracing adapter emits Chat reasoning and native Responses summaries as `llm.reasoning_content`, including when journal mode strips the full LLM message attributes.  
   **Why:** readable reasoning can remain useful after model/provider switching without teaching the target a foreign private schema. A valid summary without encrypted content is ordinary text, not a corrupt response; mixed summary-only/encrypted sequences demote together instead of partially losing reasoning. OTLP is observational, not a replay archive, so human-readable reasoning must remain available without exporting opaque protocol state.

5. **Direct wire-state stripping.**  
   **What:** raw thinking blocks, provider-specific fields, and confirmed inline signatures supplied directly in public messages are removed and reported. Only a validated NOOA envelope can restore them.  
   **Why:** otherwise callers, old middleware, or archives could bypass the compatibility gate.

6. **Visible failure policy.**  
   **What:** expected model/API incompatibility or a changed carrier warns and safely demotes text. Malformed current envelopes, unsupported known-provider shapes, ambiguous tool identity, unsupported Responses items/content blocks, invalid stored ordering, and unscopable opaque output raise `ReasoningReplayError` rather than retaining a partial turn; strategies do not retry that deterministic error.  
   **Why:** model switching is recoverable. Corrupt state or a newly changed provider contract is a framework defect and must not be silently swallowed.

7. **Trace privacy.**  
   **What:** scrubbers cover OpenAI encrypted content, Anthropic signature/data fields, Gemini snake/camel-case signatures, JSON-string attributes, and flattened inline IDs.  
   **Why:** excluding blobs from `repr` is insufficient if HTTP or tracing diagnostics can export them.

8. **No provider leakage into the event system.**  
   **What:** provider details remain concentrated in `unifiedllm/replay_state.py`; storage and rendering continue to depend only on canonical `LLMResponse`.  
   **Why:** NOOA should expose one durable turn interface, not LiteLLM objects or a catalog of provider message classes.

9. **Keep result IDs and turn bookkeeping consistent.**  
   **What:** structurally confirmed inline-signature removal also updates the matching tool result; an aborted CodeAct replay emits a final failed turn. HTTP logging reuses the shared tracing scrubber.  
   **Why:** a cleaned assistant ID must still match its result, failed generations must not look resumable, and one redaction implementation prevents provider handling from drifting between logs and traces.

The private replay envelope is now version 2 (from #310). Older version-1 draft envelopes warn and replay available plain-text reasoning, not opaque state; the stored archives themselves remain unchanged. Both OpenAI API paths validate individual reasoning items; malformed current Chat state raises `ReasoningReplayError` before dispatch.

## Supported surface

- OpenAI/Azure: Responses summaries here; encrypted OpenAI replay remains in #310
- Anthropic: Chat thinking and redacted-thinking blocks
- Gemini: Chat thought signatures, including per-tool/call-result alignment
- every provider exposing readable reasoning: generic text persistence and replay inherited from #312

Native Bedrock and Vertex opaque replay remain fail-closed until their issuer compatibility is established. NVIDIA's OpenAI-compatible routes work when the response itself carries the supported Anthropic/Gemini structure; no model-name condition is used.

## Non-goals

- no model capability catalog
- no TUI changes
- no reasoning-effort controls
- no speculative Bedrock/Vertex compatibility groups

## Validation

- scrubber/transport follow-up (2026-09-11): **71 passed** across secret scrubbing, real readable-reasoning OTLP serialization, and Responses transport override tests. Inherits #310's Unicode/prefix fix without changing provider-state replay. No inference calls.

- fingerprint/item-validation follow-up (2026-09-11): **756 passed**, 5 deselected, across UnifiedLLM and tracing. Includes 30 inherited compact-binding/validation cases, two Gemini signature-count corruption cases, and eight real OTLP serialization cases. Native Anthropic/Gemini SDK → SQLite → serialized-request probes also pass with mocked HTTP. Ruff and targeted Pyright pass; no inference calls.

- OTLP/copy-reduction round (2026-09-11): **724 passed** across tracing and UnifiedLLM; eight OTLP cases exercise real span finalization, secret scrubbing, journal-mode filtering, and HTTP JSON serialization for Chat reasoning fields, `<think>` text, and native Responses summaries. No inference calls. Ruff and targeted Pyright pass.

- latest regression round: real SDK → JSON/SQLite → serialized HTTP tests cover multi-block text, ordered/phased messages, summary-only reasoning, unsupported whole-turn rejection, and eligible Anthropic cache-marker fallback

- focused replay/formatter/tracing/strategy set after review: **152 passed**
- prior combined-stack full suite (before this follow-up): **7,364 passed**, 10 skipped, 239 deselected, 3 expected xfails
- mutations of text, tool name, arguments, IDs, order, count, state-only carriers, and direct thinking fields withhold opaque state
- Ruff, targeted Pyright, DCO trailers, and diff checks pass
- top-stack live NVIDIA Inference Hub SQLite close/reopen tests: Sonnet 5 and Gemini 3.1 Pro retained exact opaque state and stable wire requests; resumed calls reported 10,847 and 24,491 cached input tokens respectively. Sonnet 5 used the hub's native Anthropic Messages endpoint.
- additional previously verified NVIDIA Inference Hub coverage:
  - Claude Sonnet-class signed thinking survived JSON resume and an accepted follow-up
  - Gemini 3.1 Pro opaque state survived JSON resume; clean public IDs were restored to matching signed assistant/tool-result IDs and the follow-up was accepted
  - Gemini opaque state was withheld from Nemotron, which accepted the public turn

Stacked on #310.

## Summary by CodeRabbit

- **New Features**
  - Preserves portable reasoning text across supported chat and responses workflows.
  - Supports safer provider-specific reasoning-state replay, with compatibility checks and fallback behavior.
  - Exposes `ReasoningReplayError` for malformed or unsupported replay state.

- **Bug Fixes**
  - Normalizes missing tool-call arguments to empty strings.
  - Prevents replay errors from being hidden by automatic strategy retries.

- **Security**
  - Redacts additional provider-specific reasoning signatures and opaque thinking data from logs and traces.

- **Documentation**
  - Clarifies reasoning portability, replay compatibility, warnings, and error behavior.

### CODE REVIEW

File: src/nooa/unifiedllm/unifiedllm.py [hunk 1]
```diff
@@ -3012,6 +3195,7 @@ def _transform_messages(
         for original_message in messages:
             msg = copy.deepcopy(original_message)
             state = msg.pop(LLM_STATE_KEY, None)
+            reasoning = msg.pop(LLM_REASONING_KEY, None)
```
[the remark this answers is not in the document]
Comment (furgalep) [reply src/nooa/unifiedllm/unifiedllm.py]: Addressed in dd042d3e: native Responses function-call carriers now drop the private reasoning sidecar before indexing or replay, and the ordered replay regression asserts that no such key reaches input.

**Comment by furgalep** [comment]
@coderabbitai review

**Comment by furgalep** [comment]
@coderabbitai review

**Comment by furgalep** [comment]
@coderabbitai review

**Comment by furgalep** [comment]
@coderabbitai review

**Comment by furgalep** [comment]
@coderabbitai review

**Comment by furgalep** [comment]
@coderabbitai review

**Comment by furgalep** [comment]
@coderabbitai review

**Comment by furgalep** [comment]
@coderabbitai review

File: src/nooa/unifiedllm/replay_state.py [hunk 2]
```diff
@@ -187,27 +212,182 @@ def _matching_payload(state: Any, scope: str | None, state_format: str) -> dict
...
+    tool_calls = message.get("tool_calls")
+    tool_state = payload.get("tool_calls")
+    if not isinstance(tool_calls, list) or not isinstance(tool_state, list):
+        return
+    for call, state in zip(tool_calls, tool_state, strict=False):
+        if not isinstance(call, dict) or not isinstance(state, dict):
+            continue
+        fields = state.get("provider_specific_fields")
+        if isinstance(fields, dict):
+            call["provider_specific_fields"] = copy.deepcopy(fields)
```
[the remark this answers is not in the document]
Comment (furgalep) [reply src/nooa/unifiedllm/replay_state.py]: Fixed with a stricter identity gate. Capture now stores the original ordered public tool-call IDs only when opaque state exists and rejects missing/duplicate IDs. Replay compares the complete ID sequence before restoring any message-level or per-call state; dropped, reordered, renamed, added, or duplicate calls therefore retain public data but receive no opaque state. Legacy positional per-call state without IDs also fails closed. The happy path plus drop/reorder/duplicate regressions pass.

**Comment by esarafian** [comment]
Paul, I reproduced a remaining plain-text fallback gap on `cd287ba9dc294f5c03471faa0411928512a42e26`, adjacent to the [tool-call identity guard discussion](https://github.com/NVIDIA-NeMo/labs-OO-Agents/pull/311#discussion_r3972719915).

The identity guard correctly rejects opaque replay when Gemini tool-call IDs are dropped, reordered, or duplicated. However, the readable reasoning is also lost: the prepared assistant message has `content=None` despite `first.reasoning` being populated.

**Cause:** [`_restore_chat_state()`](https://github.com/NVIDIA-NeMo/labs-OO-Agents/blob/cd287ba9dc294f5c03471faa0411928512a42e26/src/nooa/unifiedllm/replay_state.py#L371-L400) returns early on an identity mismatch. [`prepare_chat_messages()`](https://github.com/NVIDIA-NeMo/labs-OO-Agents/blob/cd287ba9dc294f5c03471faa0411928512a42e26/src/nooa/unifiedllm/replay_state.py#L459-L477) chooses the restoration branch because the scoped payload exists, so it never calls `_merge_reasoning_text()` even though restoration was rejected.

**Small regression:** extend the existing parameterized `test_gemini_tool_state_fails_closed_when_public_calls_change` in `tests/unifiedllm/test_cross_provider_reasoning.py`, keeping all its signature-removal assertions, with:

```python
assert first.reasoning == "Inspect the value."
assert replayed.get("content") == first.reasoning
```

The equivalent mocked client → formatter → next-request probes fail for all three mutations (`drop`, `reorder`, `duplicate`): signatures are absent as intended, but readable reasoning is absent too. These are deterministic local tests, not a live Gemini result.

**Suggested minimal fix:** make `_restore_chat_state()` report whether restoration was accepted, and use plain-text fallback when it was rejected:

```python
# Sketch: requires the helper's new bool return contract.
restored = bool(payload) and _restore_chat_state(message, payload)
if not restored:
    _merge_reasoning_text(message, reasoning)
```

Return `False` for rejected identity/legacy positional state, before restoring anything. Treat valid message-only restoration as success even when there is no per-tool state—simply changing every existing bare `return` to `False` would incorrectly trigger fallback for that valid case. Preserve the existing issuer/identity checks and avoid partially restoring rejected state.

Alongside the three mutation cases, retain the happy-path Gemini and Anthropic signed-thinking tests to check that valid opaque replay remains unchanged and does not also duplicate reasoning into ordinary content. This suggestion is limited to selecting the existing fallback; it does not require relaxing signature validation or changing the persistence model. 🤖🤖🤖

**Comment by furgalep** [comment]
@coderabbitai review

**Comment by furgalep** [comment]
@coderabbitai review

File: src/nooa/unifiedllm/replay_state.py [hunk 3]
```diff
@@ -137,27 +143,199 @@ def _matching_payload(state: Any, scope: str | None, state_format: str) -> dict
...
 
 
+def _scope_provider(scope: str | None) -> str | None:
+    if not isinstance(scope, str):
+        return None
+    parts = scope.split(":", 2)
+    return parts[1] if len(parts) == 3 else None
+
+
+def _sanitize_chat_payload(payload: dict[str, Any], scope: str | None) -> dict[str, Any]:
```
Comment (furgalep) [inline src/nooa/unifiedllm/replay_state.py:292]: We should log a warning if we drop a payload.

File: src/nooa/unifiedllm/replay_state.py [hunk 4]
```diff
@@ -137,27 +143,199 @@ def _matching_payload(state: Any, scope: str | None, state_format: str) -> dict
...
+
+    if clean and payload.get("state_only") is True:
+        clean["state_only"] = True
+    return clean
+
+
+def _tool_call_state(tool_call: Any) -> dict[str, Any] | None:
+    dumped = opaque_item(tool_call)
+    if not isinstance(dumped, dict):
+        return None
```
Comment (furgalep) [inline src/nooa/unifiedllm/replay_state.py]: Should we log a warning here? When would this happen? Should this be a hard fail or a warning?

File: src/nooa/unifiedllm/replay_state.py [hunk 5]
```diff
@@ -137,27 +143,199 @@ def _matching_payload(state: Any, scope: str | None, state_format: str) -> dict
...
+    signature = fields.get("thought_signature") if isinstance(fields, dict) else None
+    call_id = dumped.get("id")
+    if (
+        not signature
+        and isinstance(call_id, str)
+        and _INLINE_THOUGHT_SIGNATURE_SEPARATOR in call_id
+    ):
+        signature = call_id.split(_INLINE_THOUGHT_SIGNATURE_SEPARATOR, 1)[1]
+    if not isinstance(signature, str) or not signature:
+        return None
```
Comment (furgalep) [inline src/nooa/unifiedllm/replay_state.py:397]: Same here. Is this a moment to hard fail so we don't swallow bugs?

File: src/nooa/unifiedllm/replay_state.py [hunk 6]
```diff
@@ -193,27 +214,199 @@ def _matching_payload(state: Any, scope: str | None, state_format: str) -> dict
...
+        tool_calls = message.get("tool_calls")
+        if (
+            not isinstance(tool_calls, list)
+            or [call.get("id") if isinstance(call, dict) else None for call in tool_calls]
+            != expected_call_ids
+        ):
+            return
+    elif isinstance(payload.get("tool_calls"), list):
+        # Positional per-call state cannot safely survive public call mutation.
+        return
```
[the remark this answers is not in the document]
Comment (furgalep) [reply src/nooa/unifiedllm/replay_state.py]: Addressed in 6976d773. `_restore_chat_state()` now reports acceptance; any public tool-call identity mismatch emits a warning, withholds opaque state, demotes readable reasoning, and removes an empty state-only carrier when there is no text. The drop/reorder/duplicate regression covers both state withholding and text survival.

File: src/nooa/unifiedllm/replay_state.py [hunk 7]
```diff
@@ -196,10 +395,15 @@ def prepare_chat_messages(messages: list[dict[str, Any]], scope: str | None) ->
...
+        )
         payload = _matching_payload(state, scope, _CHAT_FORMAT)
-        if payload and isinstance(payload.get("reasoning_items"), list):
-            message["reasoning_items"] = payload["reasoning_items"]
+        if payload is not None:
+            payload = _sanitize_chat_payload(payload, scope) or None
+        if payload:
+            _restore_chat_state(message, payload)
         else:
             demote_reasoning_text(message, reasoning)
```
[the remark this answers is not in the document]
Comment (furgalep) [reply src/nooa/unifiedllm/replay_state.py:823]: Addressed in 6976d773. Restoration now returns a boolean and both text demotion and empty-carrier filtering use that result, so rejected opaque state cannot suppress readable reasoning or leave an empty assistant message.

File: src/nooa/unifiedllm/replay_state.py (hunk 3 above)
Comment (furgalep) [reply src/nooa/unifiedllm/replay_state.py:292]: Addressed in 6976d773 with an explicit policy rather than another silent sanitizer branch: expected scope/carrier incompatibility warns and demotes text; unknown external provider-extension fields warn; malformed or unsupported state inside a current v1 NOOA envelope raises `ReasoningReplayError` before dispatch.

File: src/nooa/unifiedllm/replay_state.py (hunk 4 above)
Comment (furgalep) [reply src/nooa/unifiedllm/replay_state.py]: Addressed in 6976d773. A provider tool call that cannot be represented as a mapping is a known-contract violation and now hard-fails with `ReasoningReplayError`; the strategies do not retry it. A well-formed tool call with no thought signature remains a normal, silent no-state case.

File: src/nooa/unifiedllm/replay_state.py (hunk 5 above)
Comment (furgalep) [reply src/nooa/unifiedllm/replay_state.py:397]: Addressed in 6976d773. Absence of a thought signature is normal and remains silent. If the signature field is present but malformed/empty, or inline and structured signatures conflict, capture hard-fails with `ReasoningReplayError`. Unknown non-null provider extension fields warn so new state shapes do not disappear silently.

**Comment by furgalep** [comment]
@coderabbitai review

**Comment by furgalep** [comment]
@coderabbitai full review

**Comment by furgalep** [comment]
@coderabbitai full review

File: src/nooa/strategies/codeact.py [hunk 8]
```diff
@@ -919,6 +921,8 @@ async def _run_generation(
                         tool_choice=tool_choice,
                         **self._build_sampling_kwargs(),
                     )
+                except ReasoningReplayError:
+                    raise
```
[the remark this answers is not in the document]
Comment (furgalep) [reply src/nooa/strategies/codeact.py:926]: Fixed in 4d342390. ReasoningReplayError now marks the aborted CodeAct turn final before propagating. The integration test checks exactly one failed, final AfterTurn with the correct exception type for both CodeAct and PurePython.

File: src/nooa/unifiedllm/replay_state.py [hunk 9]
```diff
@@ -305,58 +522,208 @@ def _public_responses_carriers(items: list[dict[str, Any]]) -> list[dict[str, An
...
+            public_calls = cast(dict[str, Any], payload)["carrier"]["tool_calls"]
+            private_calls = message.get("tool_calls") or []
+            for public_call, private_call in zip(public_calls, private_calls, strict=True):
+                public_id = public_call["id"]
+                private_id = private_call["id"]
+                if public_id != private_id:
+                    private_call_ids[public_id] = private_id
+        tool_call_id = message.get("tool_call_id")
+        if isinstance(tool_call_id, str) and tool_call_id in private_call_ids:
+            message["tool_call_id"] = private_call_ids.pop(tool_call_id)
```
[the remark this answers is not in the document]
Comment (furgalep) [reply src/nooa/unifiedllm/replay_state.py:821]: Verified against current code; skipping the proposed reorder support. A Relay edit crosses JSON and replaces the messages, dropping their private replay sidecars, so the cited private-state-preserving reorder does not occur. Stock formatting emits calls before results; reverse chronology is invalid provider input independently of ID mapping. The review did uncover a related valid edge case: stripping a structurally identified inline signature on a raw assistant call must also strip it from its matching result. Fixed that in 4d342390 with unknown/OpenAI-target regressions and coverage preserving legitimate thought-substring IDs.

**Comment by furgalep** [comment]
Review of `d0c9b6d0`. The Anthropic/Gemini capture, the carrier fingerprint generalisation, the inline-signature split/restore, and the scrubber additions all read well and the targeted suites pass on the stack tip. Three changes requested before merge, plus a merge-strategy note.

### 1. Blocking: unscoped routes must warn and drop state, not fail the call

The "visible failure policy" in the description is right for a malformed envelope under a **known** scope. It is applied too broadly: any route that resolves to **no** scope but whose response carries provider state now raises `ReasoningReplayError` after the provider has already answered.

Concretely, `replay_scope` returns `None` for `bedrock/…`, `vertex_ai/…`, and `openrouter/…`, but LiteLLM still puts `thinking_blocks` on a Claude message from those routes, and `provider_specific_fields.thought_signature` on Gemini tool calls from Vertex. `capture_chat_state` (`src/nooa/unifiedllm/replay_state.py:417`) collects them, then `_sanitize_chat_payload` (`replay_state.py:292`) raises at `replay_state.py:362` with

```
ReasoningReplayError: Opaque chat reasoning state is malformed or unsupported for provider None.
```

`_envelope` (`replay_state.py:156`) has the same behaviour at line 161, and `capture_responses_state` (`replay_state.py:679`) raises for encrypted reasoning on a non-OpenAI scope. Reproduced end to end through `CompletionClient.call` with a mocked `bedrock/anthropic.claude-sonnet-4-*` response: the successful reply is discarded, CodeAct and PurePython re-raise, and the run dies. The existing `test_unverified_closed_provider_routes_have_no_opaque_replay_scope` (`tests/unifiedllm/test_cross_provider_reasoning.py:790`) only asserts these routes get no scope; nothing exercises capture with their state.

**Fix.** Treat `scope is None` as expected incompatibility, matching the module docstring ("expected incompatibility warns and demotes portable text"):

- In `capture_chat_state` and `capture_responses_state`, when `scope is None` and provider state is present, log one warning ("opaque reasoning state from an unsupported route was not retained; portable reasoning text will be replayed instead") and return `None`. Do not call `_sanitize_chat_payload` or `_envelope` with a `None` scope.
- Keep raising for a **known** scope with a malformed payload (unknown fields, bad fingerprints, conflicting signatures, bad `encrypted_content`). That is the corruption case the policy is meant for.
- `LLMResponse.reasoning` still carries the readable text, so cross-model replay keeps working on these routes.

**Tests.** Parametrize over `bedrock/anthropic.claude-sonnet-4-20250514-v1:0`, `vertex_ai/claude-sonnet-4`, `vertex_ai/gemini-2.5-pro`, and `openrouter/anthropic/claude-sonnet-4`. For each, drive `CompletionClient.call` with a monkeypatched `litellm.completion` returning (a) a message with `thinking_blocks`, and (b) a tool-call message with `provider_specific_fields.thought_signature` and an `__thought__` suffix on the id. Assert the call returns normally, `llm_state is None`, `reasoning` is populated where the response had text reasoning, a warning was logged, and no `ReasoningReplayError` escaped. Add a Responses variant with an `encrypted_content` reasoning item on a non-OpenAI scope. No live credentials are needed for any of this; the mocked path is exactly where the failure sits.

### 2. `responses_reasoning_text` should tolerate summary shapes

`responses_reasoning_text` (`replay_state.py:758`) runs unconditionally on every Responses reply (`unifiedllm.py:2394`, `unifiedllm.py:2528`) and again inside the OTLP patch (`_litellm_patch.py:264`). It raises if `summary` is not a list or an entry's `text` is not a string. The official OpenAI shape is a list of `{"type": "summary_text", "text": …}`, but that is not universal: LiteLLM's own reasoning-item repair builds `summary` as a plain string, and gateways that bridge Chat providers into Responses format do not guarantee the list shape. A string summary today fails the entire call, and in the tracing path it raises inside the instrumentation callback.

**Fix.** Accept a string `summary` as a single text; skip entries whose `text` is not a string with a debug log; never raise from this helper. Add tests for `summary: "text"`, `summary: null`, and an entry missing `text`.

### 3. Do not raise on an empty tool-call id

`public_tool_call_id` (`replay_state.py:406`) raises at line 410 when a tool call has no id. Before this PR an empty id flowed through as `""`, which is what some OpenAI-compatible NIM and vLLM endpoints emit. Those models have no opaque state, so there is nothing to protect.

**Fix.** Return `""` for a missing or empty id, as before. In `capture_chat_state`, if opaque state is present **and** the carrier cannot be built (empty or duplicate ids), warn and return `None` rather than raise; the fingerprint cannot bind state to an unidentifiable call, so demoting to text is the correct fallback. Please do not synthesise ids here: `_chat_public_carrier` fingerprints the raw message at capture, and a generated id would not match the id rendered back from the event, so the state would never replay. Add a test with a tool call whose `id` is `""` on a plain OpenAI-compatible route asserting the `ToolCall` is produced and nothing raises.

### 4. Squash on merge

The branch carries several stack-sync merge commits ("Merge compact replay bindings…", "Merge whole-turn replay validation…"). Please squash-merge so `main` stays linear.

**Comment by furgalep** [comment]
Superseded by #318, which includes Anthropic/Gemini native state and portable plain-text reasoning in ordered response parts. Callers return LLMResponse objects directly to UnifiedLLM, removing the reconstruction and private-carrier machinery. Cache policy follows in #319. This branch and its review history are retained.
